### PR TITLE
[DOCS] More on reference types.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: docs/types.rst [hunk 1]
```diff
@@ -847,15 +847,34 @@ Members
         // dynamic array of pairs (i.e. of fixed size arrays of length two).
```
Comment (chriseth) [inline docs/types.rst:942]: This example needs some more love, but probably better to do this after the 0.5.0 release.

**Review by leonardoalt — DISMISSED** [review]

File: docs/types.rst [hunk 2]
```diff
@@ -728,9 +722,13 @@ third dynamic array, you use ``x[2][1]`` (indices are zero-based and
...
 
+Accessing an array past its end will cause a revert. If you want to add
+new elements, you have to use ``.push()`` or increase the ``.length``
+member (see below).
+
 Variables of type ``bytes`` and ``string`` are special arrays. A ``bytes`` is similar to ``byte[]``,
-but it is packed tightly in calldata. ``string`` is equal to ``bytes`` but does not allow
-length or index access (for now).
+but it is packed tightly in calldata and memory. ``string`` is equal to ``bytes`` but does not allow
+length or index access.
```
Comment (axic) [inline docs/types.rst:813]: index access or indexed access?

File: docs/types.rst [hunk 3]
```diff
@@ -823,13 +821,25 @@ Members
     Arrays have a ``length`` member to hold their number of elements.
     Dynamic arrays can be resized in storage (not in memory) by changing the
     ``.length`` member. This does not happen automatically when attempting to access elements outside the current length. The size of memory arrays is fixed (but dynamic, i.e. it can depend on runtime parameters) once they are created.
+    Increasing the length of an array creates new zero-initialized elements while decreasing the length
+    deletes the members (sets them to zero).
```
Comment (axic) [inline docs/types.rst]: How about merging #4865 first and then rebase this PR?

**Comment by axic** [comment]
I think this applies the same changes as #4865, but in a different form. Please compare the two and make a decision which approach to take.

File: docs/types.rst (hunk 2 above)
Comment (chriseth) [reply docs/types.rst:813]: not sure, I would call it `index access`

**Comment by chriseth** [comment]
Tried to combine ideas from both PRs.

File: docs/types.rst [hunk 4]
```diff
@@ -674,14 +674,14 @@ non-persistent area where function arguments are stored, and behaves mostly like
...
-always create an independent copy.
-Assignments to local storage variables only assign a reference though, and
-this reference always points to the state variable even if the latter is changed
-in the meantime.
-On the other hand, assignments from a memory stored reference type to another
-memory-stored reference type do not create a copy.
+Data locations are not only relevant for persistency of data, but also for the semantics of assignments:
+assignments between storage and memory (or from calldata) always create an independent copy.
+Assignments from memory to memory only create references. This means that changes to one memory variable
+will also be visible in all other memory variables that refer to the same data.
```
Comment (ChrisChinchilla) [inline docs/types.rst]: "… are also visible…"

File: docs/types.rst [hunk 5]
```diff
@@ -674,14 +674,14 @@ non-persistent area where function arguments are stored, and behaves mostly like
...
+Assignments from storage to a local storage variables also only assign a reference.
+In contrast, all other assignments to storage will always copy. Examples for this case
```
Comment (ChrisChinchilla) [inline docs/types.rst]: remove "will".

File: docs/types.rst [hunk 6]
```diff
@@ -726,9 +719,10 @@ Arrays
 ------
 
 Arrays can have a compile-time fixed size or they can be dynamic.
-For storage arrays, the element type can be arbitrary (i.e. also other
-arrays, mappings or structs). For memory arrays, it cannot be a mapping and
-has to be an ABI type if it is an argument of a publicly-visible function.
+The are very few restrictions for the element, it can also be
```
Comment (ChrisChinchilla) [inline docs/types.rst]: Remove "very"

File: docs/types.rst [hunk 7]
```diff
@@ -726,9 +719,10 @@ Arrays
...
+another array, a mapping or a struct. The general restrictions for
+types apply, though, in that mappings can only be used in storage
```
Comment (ChrisChinchilla) [inline docs/types.rst:796]: "… you can only use mappings in storage…"

File: docs/types.rst [hunk 8]
```diff
@@ -738,9 +732,13 @@ third dynamic array, you use ``x[2][1]`` (indices are zero-based and
 access works in the opposite way of the declaration, i.e. ``x[2]``
 shaves off one level in the type from the right).
 
+Accessing an array past its end will cause a revert. If you want to add
```
Comment (ChrisChinchilla) [inline docs/types.rst]: "…its end causes a revert…"

File: docs/types.rst [hunk 9]
```diff
@@ -760,7 +758,7 @@ The numeric index will become a required parameter for the getter.
 Allocating Memory Arrays
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
-Creating arrays with variable length in memory can be done using the ``new`` keyword.
+Creating arrays with a runtime-dependent length in memory can be done using the ``new`` keyword.
```
Comment (ChrisChinchilla) [inline docs/types.rst]: "You can create arrays with a runtime-dependent length in memory using the…"

File: docs/types.rst [hunk 10]
```diff
@@ -830,15 +828,30 @@ Members
 ^^^^^^^
 
 **length**:
-    Arrays have a ``length`` member to read their number of elements.
-    Dynamically-sized arrays (only available for storage) have a read-write ``length`` member to resize the array. Increasing the length adds uninitialized elements to the array, this has *O(1)* complexity. Reducing the length performs :ref:``delete`` on each removed element and has *O(n)* complexity where *n* is the number of elements being deleted. Please note that calling ``length--`` on an empty array will set the length of the array to 2^256-1 due to ``uint256`` underflow wrapping.
+    Arrays have a ``length`` member that contains their number of elements.
+    The length of memory arrays is fixed (but dynamic, i.e. it can depend on runtime parameters) once they are created.
+    For dynamically-sized arrays (only available for storage), this member can be assigned to resize the array.
+    Accessing elements outside the current length will not automatically resize the array and instead cause a failing assertion.
```
Comment (ChrisChinchilla) [inline docs/types.rst]: "…does not automatically…"

Comment (ChrisChinchilla) [inline docs/types.rst]: "…instead causes…"

File: docs/types.rst [hunk 11]
```diff
@@ -830,15 +828,30 @@ Members
...
+    Increasing the length adds new zero-initialized elements to the array.
+    Reducing the length performs an implicit :ref:``delete`` on each of the removed elements.
 **push**:
-     Dynamic storage arrays and ``bytes`` (not ``string``) have a member function called ``push`` that can be used to append an element at the end of the array. The function returns the new length.
+     Dynamic storage arrays and ``bytes`` (not ``string``) have a member function called ``push`` that can be used to append an element at the end of the array. The element will be zero-initialized. The function returns the new length.
```
Comment (ChrisChinchilla) [inline docs/types.rst]: "…that you can use…" and "The element is zero-initialised".

File: docs/types.rst [hunk 12]
```diff
@@ -830,15 +828,30 @@ Members
...
 **pop**:
      Dynamic storage arrays and ``bytes`` (not ``string``) have a member function called ``pop`` that can be used to remove an element from the end of the array. This will also implicitly call :ref:``delete`` on the removed element.
 
 .. warning::
-    It is not yet possible to use arrays of arrays in external functions.
+    If you use ``.length--`` on an empty array, it will cause an underflow and
```
Comment (ChrisChinchilla) [inline docs/types.rst]: "…it causes an underflow…"

File: docs/types.rst [hunk 13]
```diff
@@ -830,15 +828,30 @@ Members
...
+    thus sets the length to ``2**256-1``.
+
+.. note::
+    Increasing the length of a storage array has constant gas costs because
+    storage is assumed to be zero-initialized, while decreasing
```
Comment (ChrisChinchilla) [inline docs/types.rst]: zero-initialised

**Review by ChrisChinchilla — CHANGES_REQUESTED** [review]
@chriseth Some good changes, just some minor grammar and spelling recommendations for reading clarity.

File: docs/types.rst (hunk 7 above)
Comment (chriseth) [reply docs/types.rst:796]: I would like to keep it like that to emphasize that it means that mappings can only be used in storage as opposed to mappings are the only types that can be used in storage.

Comment (ChrisChinchilla) [reply docs/types.rst:796]: @chriseth Ahh, now I wonder if there may even be better ways to clarify this. Maybe (and feel free to paraphrase):

"You can only store/use mappings in the storage data location" (I believe that's the term we started to use)

Or, keeping it passive, but maybe a bit wordy:

"mappings can only be stored in the storage data location"

Comment (chriseth) [reply docs/types.rst:796]: I would use "storage data location" only for the data location specifier, and not the general term. For example in `contract c { mapping(uint => uint) x; }` the mapping is in storage but there is no explicit data location specifier.

Comment (ChrisChinchilla) [reply docs/types.rst:796]: OK, not to worry, whatever you feel is best from your initial idea, my ideas, or a combination of.

**Review by leonardoalt — DISMISSED** [review]

File: docs/types.rst [hunk 14]
```diff
@@ -836,15 +834,30 @@ Members
...
 
 .. warning::
-    It is not yet possible to use arrays of arrays in external functions.
+    If you use ``.length--`` on an empty array, it causes an underflow and
+    thus sets the length to ``2**256-1``.
+
+.. note::
+    Increasing the length of a storage array has constant gas costs because
+    storage is assumed to be zero-initialised, while decreasing
+    the length has at least linear cost, because it includes explicitly clearing the removed
```
Comment (axic) [inline docs/types.rst]: Maybe emphasize `has at least linear cost (but in many cases worse than linear)` ?

File: docs/types.rst [hunk 15]
```diff
@@ -862,15 +875,34 @@ Members
...
             m_pairsOfFlags = newPairs;
         }
 
+        struct StructType {
+            uint[] contents;
+            uint moreInfo;
+        }
+        StructType s;
+
+        function f(uint[] memory c) public {
```
Comment (axic) [inline docs/types.rst:962]: Is this example still compiling with 0.4.16 (that is the pragma above) ?

File: docs/types.rst [hunk 16]
```diff
@@ -1010,16 +1046,19 @@ or complex types like contract types, enums, mappings, structs and any array typ
...
 such that every possible key exists and is mapped to a value whose
 byte-representation is all zeros, a type's :ref:`default value <default-value>`. The similarity ends there, the key data is not stored in a
 mapping, only its ``keccak256`` hash is used to look up the value.
 
 Because of this, mappings do not have a length or a concept of a key or
 value being set.
 
-Mappings are **only** allowed for state variables (or as storage reference types
-in internal functions).
+Mappings can only have a data location of "storage" and thus
```
Comment (axic) [inline docs/types.rst]: Why not ` ``storage`` ` ? The documentation is inconsistent about this though.

File: docs/types.rst (hunk 15 above)
Comment (chriseth) [reply docs/types.rst:962]: confirmed.

**Comment by chriseth** [comment]
Updated.

**Review by axic — DISMISSED** [review]
Please squash first.

File: docs/types.rst [hunk 17]
```diff
@@ -1022,11 +1058,12 @@ Struct types can be used inside mappings and arrays and they can itself
...
 
 It is not possible for a struct to contain a member of its own type,
-although the struct itself can be the value type of a mapping member.
+although the struct itself can be the value type of a mapping member
+or it can contain a dynamically-sized array of its type.
 This restriction is necessary, as the size of the struct has to be finite.
 
 Note how in all the functions, a struct type is assigned to a local variable
-(of the default storage data location).
+with data location storage.
```
Comment (erak) [inline docs/types.rst]: Maybe use ``storage``?

File: docs/types.rst [hunk 18]
```diff
@@ -566,7 +566,7 @@ also accepts a payment of zero Ether, so it also is ``non-payable``.
 On the other hand, a ``non-payable`` function will reject Ether sent to it,
 so ``non-payable`` functions cannot be converted to ``payable`` functions.
 
-If a function type variable is not initialized, calling it will result
+If a function type variable is not initialised, calling it will result
```
Comment (erak) [inline docs/types.rst]: Perhaps remove future tense here?

**Review by erak — DISMISSED** [review]
Just found to minor things that we might want to change 😉

File: docs/types.rst (hunk 18 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: @bit-shift It's a good call, as it happens immediately in this case. The `will` above doesn't, as far as I can understand anyway.

File: docs/types.rst (hunk 17 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: @bit-shift As in the formatting?

**Comment by chriseth** [comment]
Fixed, rebased, squashed.
